invenio_records_rest/__init__.py

```diff
@@ -22,7 +22,649 @@
+
+
+and retrieve the citation for it with:
+
+.. code-block:: console
+
+    $ curl -H "Accept:text/x-bibliography" -XGET <HOST>/records/<PID>
+    First Name, Last Name, Sample Record.
+
+
+In a similar way, schemas can be defined to be used with a serializer,
+inheriting from the ``MarshmallowMixin``, and do any type of transformation
+to the JSON objects that are exchanged in the REST API.
+
+Data formats
+~~~~~~~~~~~~
+Typical serialization formats are:
+
+- JSON: JSON-LD, CSL
+- XML: DataCite, DublinCore, MARCXML
+- Text: Citation formatting
+
+For some examples on the implementation of these serializers you can see
+the ones present in different invenio instances such as:
+
+- `CDS-Videos <https://github.com/CERNDocumentServer/cds-videos/tree/cdslabs_qa/cds/modules/records/serializers>`_
+- `Zenodo <https://github.com/zenodo/zenodo/tree/master/zenodo/modules/records/serializers>`_
+
+
+Tombstones and redirection
+--------------------------
+
+When it comes to manage changes regarding PIDs such as deletions and
+redirections, Invenio-Records-REST offers ways to:
+
+* Customize tombstones:
+
+>>> from invenio_records_rest.errors import PIDDeletedRESTError
+>>> def custom_deleted_pid_error_handler(error):
+...     record = error.pid_error.record or {}
+...     description={
+...         'status': 410,
+...         'message': error.description,
+...         'removal_reason': record.get('removal_reason')}
+...     return PIDDeletedRESTError(description=description).get_response()
+...
+>>> app.config.update({'RECORDS_REST_ENDPOINTS':
+...   {'recid':
+...      {'error_handlers':
+...         {'PIDDeletedRESTError': custom_deleted_pid_error_handler}
+...      }
+...   }
+... })
+
+* Manage redirects:
+
+ (reference with PIDstore, explain how it scans RECORDS_REST_ENDPOINTS to pick
+  the target endpoint)
+
+Access control
+--------------
+
+Invenio-Records-REST is highly customizable in terms of access control. This
+customization can be done by using permission factories. By default two
+permission factories are offered within the package, ``allow_all`` and
+``deny_all``. For example, if we wanted to allow creation and retrieval
+of records by any user, and deny deletion and update for all users, we would
+to the following:
+
+>>> from invenio_records_rest.utils import allow_all, deny_all
+>>> RECORDS_REST_ENDPOINTS = dict(
+...    recid=dict(
+...        create_permission_factory_imp=allow_all,
+...        delete_permission_factory_imp=deny_all,
+...        update_permission_factory_imp=deny_all,
+...        read_permission_factory_imp=allow_all,
+...    )
+... )
+
+Note: In order to have a full setup with access controll you should install
+`Invenio-Accounts <https://invenio-accounts.readthedocs.io/en/latest/>`_.
+
+It is also possible to filter search results based on the status of the
```

Comment:
The permissions are actually affecting the `<prefix>_item` endpoint and not the `<prefix>_list` (aka "search") endpoint, so this protects `/records/1234` from someone that doesn't have permissions to access it but it doesn't protect `/records` from returning results containing record `1234`.

In order to filter on the level of the search results (e.g. for deposit search results), one has to do the filtering via  [customizing the `RecordsSearch` class that is used](https://github.com/inveniosoftware/invenio-deposit/blob/f243ea1d01ab0a3bc92ade3262d1abdd2bc32447/invenio_deposit/search.py#L36-L68) and passing it in in the [`search_class` field of the endpoint's config](https://github.com/inveniosoftware/invenio-deposit/blob/f243ea1d01ab0a3bc92ade3262d1abdd2bc32447/invenio_deposit/config.py#L70)